File: 15.三数之和.js
/*
 * @lc app=leetcode.cn id=15 lang=javascript
 *
 * [15] 三数之和
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    nums.sort((a, b) => a - b);
    let ret = [];
    for (let i = 0; i < nums.length; i++) {
        if (i > 0 && nums[i - 1] === nums[i]) {
            continue;
        }
        let k = nums.length - 1;
        for (let j = i + 1; j < nums.length; j++) {
            if (j > i + 1 && nums[j - 1] === nums[j]) {
                continue;
            }

            while (j < k && nums[i] + nums[j] + nums[k] > 0) {
                k--;
            }

            if (j === k) {
                break;
            }

            if (nums[i] + nums[j] + nums[k] === 0) {
                ret.push([nums[i], nums[j], nums[k]]);
            }
        }
    }
    return ret;
};
// @lc code=end

File: 15.三数之和.rs
/*
 * @lc app=leetcode.cn id=15 lang=rust
 *
 * [15] 三数之和
 */

// @lc code=start
impl Solution {
    pub fn three_sum(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        if nums.len() < 3 {
            return vec![];
        }
        nums.sort_unstable();
        let mut ret = vec![];
        let mut lastnum = 1;
        for (i, &num) in nums.iter().enumerate() {
            if num > 0 {
                break;
            }
            if num == lastnum {
                continue;
            }
            let mut l = i+1;
            let mut r = nums.len() - 1;
            while l < r {
                let sum = num + nums[l] + nums[r];
                if sum == 0 {
                    ret.push(vec![num, nums[l], nums[r]]);
                    while l < r && nums[l] == nums[l + 1] {
                        l += 1;
                    }
                    l += 1;
                } else if sum < 0 {
                    l += 1;
                } else {
                    r -= 1;
                }
            }
            lastnum = num;
        }
        ret
    }
}
// @lc code=end